src/claimtrie.cpp

```diff
@@ -137,6 +156,109 @@ void CClaimTrieNode::reorderClaims(supportMapEntryType& supports)
     std::make_heap(claims.begin(), claims.end());
 }
 
+// trieCache is an input parameter on enable; an output parameter on disable.
+CClaimTrie* CClaimTrie::enableNormalizationFork(bool enable, int nHeight, CClaimTrieCache& trieCache)
+{
+    LogPrintf("%s: Normalize Name Fork is %s\n", __func__, (enable ? "ENABLED" : "DISABLED"));
+
+    if (enable) {
+        if (fNormalize) {
+            return this;
+        }
+
+        std::vector<namedNodeType> nodes = flattenTrie();
+        LogPrintf("Pre-normalization fork claim trie nodes = %lu\n", nodes.size());
+
+        if (normalizationForkStateInstance.nonNormalizedTrie == NULL) {
+            normalizationForkStateInstance.nonNormalizedTrie = this;
+        }
+
+        if (normalizationForkStateInstance.nonNormalizedTrieCache == NULL) {
+            normalizationForkStateInstance.nonNormalizedTrieCache = boost::make_shared<CClaimTrieCache>(this);
+            normalizationForkStateInstance.nonNormalizedTrieCache->cloneState(trieCache);
+            LogPrintf("%s: Cloned Merkle hash %s, Original %s\n", __func__, normalizationForkStateInstance.nonNormalizedTrieCache->getMerkleHash(true).GetHex().c_str(), trieCache.getMerkleHash(true).GetHex().c_str());
+        }
+
+        if (normalizationForkStateInstance.normalizedTrie == NULL) {
+            normalizationForkStateInstance.normalizedTrie = new CClaimTrie(true, false, true);
+            assert(normalizationForkStateInstance.normalizedTrie->shouldNormalize());
+            normalizationForkStateInstance.nonNormalizedTrie->WriteToDisk();
+            normalizationForkStateInstance.normalizedTrie->ReadFromDisk(true);
+            normalizationForkStateInstance.normalizedTrie->nCurrentHeight = nHeight;
+            normalizationForkStateInstance.normalizedTrie->copyAndNormalizeQueues(*this);
+
+            nodes = normalizationForkStateInstance.normalizedTrie->flattenTrie();
+            LogPrintf("Post-normalization fork claim trie nodes = %lu\n", nodes.size());
+        }
+
+        assert(normalizationForkStateInstance.nonNormalizedTrie);
+        return normalizationForkStateInstance.normalizedTrie;
+    }
+
+    if (normalizationForkStateInstance.nonNormalizedTrieCache != NULL) {
+        assert(normalizationForkStateInstance.nonNormalizedTrieCache != NULL);
+        LogPrintf("%s: Restoring cloned Merkle hash %s, Empty %s\n", __func__, normalizationForkStateInstance.nonNormalizedTrieCache->getMerkleHash(true).GetHex().c_str(), trieCache.getMerkleHash(true).GetHex().c_str());
```

Comment:
Again, regenerating the hashes for the entire trie (x2) is too expensive for logging.